### DISCUSSION TITLE
Request duplication in Narwhal and MirBFT/ISS

### DISCUSSION THREAD

**Discussion by matejpavlovic** [consensus-shipyard/consensuslab#93] [Open problems]
The reader is assumed to be familiar with [Narwhal](https://arxiv.org/abs/2105.11827) (as a mempool) and the [ISS consensus/TOB protocol](https://arxiv.org/abs/2203.05681). This discussion is addressing the problem of transaction/request duplication in these two systems. While one (Narwhal) is subject to duplication, the other (ISS) is not. The way request duplication is prevented in ISS, however, makes ISS vulnerable to a mobile adversary (see also #91). In particular, we discuss

1. a proposed modification of the Narwhal protocol that could decrease the number of duplicate proposals and
2. whether a duplication is still a relevant issue even if a smarter (pull-based) approach is used to disseminate transaction/request payloads among nodes.

## The Problem

In Narwhal, nodes receive transactions from clients, create blocks of those transactions and broadcast the blocks to other nodes. When a client sends its transaction to multiple nodes (even an honest client might be forced to do so in a Byzantine environment), multiple nodes will broadcast the same transaction in their respective batches, wasting resources.

MirBFT and ISS (focusing henceforth on ISS) are designed to prevent such duplication through partitioning the space of all possible requests (transactions in Narwhal terminology) into so-called *buckets* and making sure only one node at a time will be able to broadcast a request. This prevents wasting resources, but is vulnerable to a mobile adversary that can always prevent the node currently in charge of a particular request from broadcasting that request (see #91)

## 1. (Partial) Deduplication in Narwhal

Given that Narwhal works in rounds, a bucket-based deduplication inspired by ISS could potentially also be used in Narwhal as follows:
- Partition the transaction space in buckets, same as in ISS
- Divide the sequence of all rounds in finite epochs, each epoch consisting of *k* consecutive rounds.
- In a given epoch, only allow a subset of nodes (in the extreme case a single node) to broadcast transactions from a given bucket. (This is simply achieved by considering a block invalid if it contains a request not belonging to the correct bucket.)
- Rotate the assignment of buckets to nodes across epochs.

### Perofrmance intuition

First thing to note is that, unlike in Mir/ISS, this approach does not completely prevent request duplication. The asynchronous nature of Narwhal does not allow for synchronization and "clean" bucket handover at epoch transitions the same way as it occurs in ISS. A node that falls behind can broadcast a transaction in an old epoch concurrently with a node that already progressed to a new epoch.

However, by the design of Narwhal, at most f nodes can fall behind this way. Moreover, in practice, wasteful duplication is only expected to occur at epoch boundaries by those correct nodes whose certificates were not considered for a round transition at epoch boundary. If a node is malicious or it falls behind severely (e.g. due to a remote geographic location), we cannot count on its resources to contribute much to overall system performance anyway.

### Liveness argument

The above approach seems to be live if the client (eventually, in the worst case) sends its transaction to 2f+1 nodes (maybe f+1 is enough, but it seems less obvious). The latency might substantially grow in the very worst case, but eventually every transaction will be included in the DAG of every correct node.

Let *tx* be the oldest transaction present at f+1 correct nodes (we assume nodes broadcast received transactions in FIFO order). At latest after n = 3f + 1 epochs, all f+1 correct nodes will have, at some point, broadcast a block with *tx*. Thus, among the 2f+1 certificates gathered by any correct node to advance to the firs round of the 3f+2nd epoch, there will be at least one (transitive) reference to a block containing *tx*.

#### Question:

Can it work? (Or does it at least help?)

## 2. Taking a Step Back

Given a well engineered implementation of transaction ("request" in ISS terms) payload dissemination in both Narwhal and ISS, what is the importance of deduplication? Both Narwal and (the currently envisioned implementation of) ISS are designed to avoid wasting network resources on transmitting transaction payload over the network. Instead using a push-based approach to broadcasting transactions, they only exchange references (hashes) of the transactions and pull the payload as needed.

If, even using the pull-based approach, payload transmission remains the main performance bottleneck, deduplication at protocol level becomes an unnecessary complication, adding useless complexity to the protocol. However, if the duplicate processing of transaction references can act as a bottleneck, deduplication at protocol level is still relevant.
Experimental evaluation could shed more light on this issue.

#### Question:

Dow we even need to de-duplicate the references, if we pull (and not push) the payloads?

**Comment by vukolic** [comment]
I read the description but I still lack the frame a bit. What is the main claim/tradeoff we discuss here?

**Reply by matejpavlovic** [reply]
1. I'm proposing a way to mitigate duplication in Narwhal. It's just a coarse idea that I'm putting in place here, and I'm wondering whether something like that can actually work and achieve the desired goal (decreasing request duplication in Narwhal). In short, my question is: Could that work?
2. I'm questioning the very necessity of fighting request duplication, in particular in the case where only (potentially highly duplicated) *references* to transactions/requests are exchanged among the nodes in the protocol itself, and the nodes "lazily" pull the payloads in a more efficient way. In short, my question is: Do we even need to de-duplicate?

I updated the initial post to reflect this answer and clarify a bit.

**Comment by LefKok** [comment]
I think the approach works, if the client has knowledge of the bucketing then it can even minimize transmissions by contacting only the next epoch leaders. There is asynchrony at interplay here that requires 2f+1 such that at least one honest makes it in the DAG (f dishonest ignore the request and f honest are slowed down). 

A nice property is that censorship resistance in this case is tied to liveness since the DAG cannot progress at all without allowing the transaction to enter the causal history of the next leaders. Hence we get non-zero chain-quality under asynchrony (which is a first since so far it is believed that chain-quality is tied to synchrony!). 

As for the point on deduplication, I see your point that in the worse case we will anyway duplicate. But in the common case deduplication makes room for better goodput so it might be worth it.  
A final note tying to #92, duplication is not always a bad thing. In fact it is the main strategy towards order-fairness hence I would want to study if we actually want to allow full duplication for some transactions that need ordering guarantees (and charge higher fees)

**Reply by matejpavlovic** [reply]
Great! Yes it's right, it seems like it doesn't work if the client only sends to f+1, unless we somehow make sure that *every* honest block eventually somehow makes it in the final DAG, but that's probably now worth it.

For the last point, yes in the worst case we will end up duplicating anyway, but I didn't mean that as my main point. I also think that for the common case it is probably worth it. What I meant is that, especially if we (optimistically) de-duplicate Narwhal, another level of de-duplication at the "total order layer" level (e.g. if we couple Narwhal as a mempool with a separate consensus protocol that orders references to blocks) might not be necessary, if that total order layer is not the system bottleneck. We could even use a multi-leader protocol **without** de-duplication (i.e. with duplication) on purpose, as you mentioned. If the total order layer is not a bottleneck (i.e. is underutilized) and we "over-provision" it by introducing redundancy in the proposals to compensate for faulty nodes, in addition to affecting order-fairness (as you said), it might significantly improve latency in some failure cases. In general, optimizing for latency instead of throughput at the total order layer might be a good idea.

**Reply by LefKok** [reply]
Ah! yes, I totally agree with your point there. If we deduplicate at Narwhal then deduplicating tiny hashes for a steep overhead seems an overkill. 

One extra dimension we need to consider if using narwhal is code complexity. Narwhal has already two embedded consensus mechanisms (Tusk and Bullshark) each of which is <300 LOC. Interfacing it with a full TOB layer will add probably 2-5k LOC
so we need to think carefully about the tradeoffs. 

Also as a side comment, with Bullshark we show how to achieve bounded time/storage validity after GST using median timestamping, hence your first comment on only needing f+1 proposers per transaction would hold after GST ;).

**Comment by vukolic** [comment]
What would be the network timing model for total order that we would consider? Regardless of that, agressive all to all communication TO protocols have the lowest latency, and it seems e.g., Tusk might be suboptimal if latency is a primary metric for TO.

**Reply by LefKok** [reply]
Yes, Tusk is suboptimal but withstands DDoS attacks. 

To get to optimal with DAG we should look into Bullshark (https://arxiv.org/abs/2201.05677 , the new version will be out soon).
We have two versions there one with a fallback that withstands DDoS attacks (but it is complex to implement) and a super simple partial synchronous. 
Without DAG, Jolteon (or the DDoS resilient version Ditto) (https://arxiv.org/abs/2106.10362) reach the minimum number of round trips to commit (we could enhance them with Zyzzyvva style fast commit, but unclear how often this actually helps versus the extra complexity).
